### PR TITLE
Add filter to allow modification of report columns

### LINKED ISSUE #4147: Filters to alter the columns used by a DataStore class
Description of the issue
--
I need to be able to extend the reports generated by the WC Admin so that it takes different values. This is needed to be able to prepare correct summary reports in a multi-currency scenario.

Current behaviour
DataStore uses columns like the following (only relevant columns included):
~~~
$this->report_columns = array(
	'gross_sales'     => $gross_sales,
	'total_sales'     => "SUM({$table_name}.total_sales) AS total_sales",
	'coupons'         => 'COALESCE( SUM(discount_amount), 0 ) AS coupons', // SUM() all nulls gives null.
	'refunds'         => "{$refunds} AS refunds",
	'taxes'            => "SUM({$table_name}.tax_total) AS taxes",
	'shipping'        => "SUM({$table_name}.shipping_total) AS shipping",
	'net_revenue'     => "SUM({$table_name}.net_total) AS net_revenue",
	'avg_order_value' => "SUM( {$table_name}.net_total ) / SUM( CASE WHEN {$table_name}.parent_id = 0 THEN 1 ELSE 0 END ) AS avg_order_value",
);
~~~
This produces incorrect results when orders are in different currencies. For example, in a store with EUR as the base currency, running `SUM({$table_name}.total_sales) AS total_sales` against orders for 100 EUR, 100 USD and 100 GBP, would return a total 300 EUR. That's obviously incorrect.

A relatively easy way to fix this issue would be to alter the columns returned by the SELECT, as follows:
~~~
'total_sales' => "SUM({$table_name}.total_sales* FX_RATE_META.meta_value) AS total_sales",
'coupons' => 'COALESCE( SUM(discount_amount * FX_RATE_META.meta_value), 0 ) AS coupons', // SUM() all nulls gives null.
'refunds' => "{$refunds} AS refunds",
'taxes' => "SUM({$table_name}.tax_tota * FX_RATE_META.meta_value) AS taxes",
'shipping' => "SUM({$table_name}.shipping_total * FX_RATE_META.meta_value) AS shipping",
'net_revenue' => "SUM({$table_name}.net_total * FX_RATE_META.meta_value) AS net_revenue",
'avg_order_value' => "SUM( {$table_name}.net_total * FX_RATE_META.meta_value ) / SUM( CASE WHEN {$table_name}.parent_id = 0 THEN 1 ELSE 0 END ) AS avg_order_value",
~~~
Note the addition of `FX_RATE_META.meta_value` to convert the amounts to the same currency. The value of `FX_RATE_META.meta_value` would be the exchange rate stored by the multi-currency plugin (preferably mine, Aelia FTW 😁), used to convert each value to its shop's base currency equivalent. With this trick, the values mentioned before would change as follows:
* 100 USD -> 92 EUR
* 100 EUR -> 100 EUR
* 100 GBP -> 115.11 EUR
* **Total**: 307.11 EUR, which is the correct value

Proposed solution
--
Adding a filter to the return value of `DataStore::assign_report_columns()` should be sufficient to allow filtering the columns with the new ones, adding the necessary calculations.

Here's an example, applied to class ` Automattic\WooCommerce\Admin\API\Reports\Order\DataStore`:
~~~
protected function assign_report_columns() {
	$table_name = self::get_db_table_name();
	// Avoid ambigious columns in SQL query. 
	$this->report_columns = apply_filters('wc_analytics_orders_report_columns', array(
		'order_id'         => "{$table_name}.order_id",
		'parent_id'        => "{$table_name}.parent_id",
		'date_created'     => "{$table_name}.date_created",
		'date_created_gmt' => "{$table_name}.date_created_gmt",
		'status'           => "REPLACE({$table_name}.status, 'wc-', '') as status",
		'customer_id'      => "{$table_name}.customer_id",
		'net_total'        => "{$table_name}.net_total",
		'total_sales'      => "{$table_name}.total_sales",
		'num_items_sold'   => "{$table_name}.num_items_sold",
		'customer_type'    => "(CASE WHEN {$table_name}.returning_customer = 1 THEN 'returning' WHEN {$table_name}.returning_customer = 0 THEN 'new' ELSE '' END) as customer_type",
	), $table_name, $this);
}
~~~
Then, in my custom code, I write a filter for `wc_analytics_orders_report_columns` and replace the appropriate one. Example:
~~~
public function wc_analytics_orders_report_columns($columns, $table_name, $data_store) {
	// ORDERS_BCER stays for ORDERS_BASE_CURRENCY_EXCHANGE_RATE. It's an alias for
	// the post_meta table, where the exchange rate is stored
	$columns['net_total'] = "({$table_name}.net_total * ORDERS_BCER.meta_value)";
	$columns['total_sales'] = "({$table_name}.total_sales * ORDERS_BCER.meta_value)";
	return $columns;
}
~~~
With this trick, the totals become correct.

**Notes**
* In my example, I add the filter as `wc_analytics_orders_report_columns`, just to have a name to use for the testing. The actual filter could have a different name, such as `woocommerce_analytics_datastore_{$context}_report_columns`. The important aspect is that such filter is added to all assign_report_columns() methods. Any report that deals with order totals must be extended, hence the need for the filter.

Alternatives considered
--
There isn't any, as far as I could see. 🤷‍♂️

Should this be prioritized?
--
Kind of. Multi-currency sites are becoming ubiquitous and the Analytics still don't allow to handle them properly. Also, adding these filters can take some time, but it shouldn't be too difficult.

Additional context
--
The proposed solution relies on a strict assumption: all the tables used by the WC Admin tool must **NOT** store "global" aggregate data, but only order-specific data (i.e. that there is **ALWAYS** one row per order). That's absolutely vital. If there were any table with an aggregate figure, for example containing the sum of multiple orders (e.g. the "300" from my initial example), that would make it impossible to calculate correct the figures. That's because each order can have a different exchange rate, which can't be applied to a single total.

If any "summary" table is going to be introduced in the future, the multi-currency aspect must be taken into account before such feature is implemented. I will be happy to discuss this with your developers, so that the common pitfalls can be avoided. I have quite a bit of experience in this field (I'm the one who brought multi-currency to WooCommerce, 7 years ago), I know what lurks around the corner. 🙂

### PR DESCRIPTION
**Pull request by joshuatf**
Fixes #4147

Adds a filter to all report columns to allow adding or modifying existing columns at the database level.

@daigo75 Could you let me know if this adequately covers your use case?

### Detailed test instructions:

Note that you may need to flush your transients cache if the report columns aren't updated in the response.

1. Add a filter and add or change existing columns:
```php
		add_filter( 'woocommerce_admin_report_columns', function( $columns, $context, $table_name ) {
			if ( 'orders' === $context ) {
				$columns['my_column'] = "RAND(2) as my_column";
				$columns['net_total'] = "({$table_name}.net_total * 2) as net_total";
			}
			return $columns;
		}, 10, 3 );
```
2. Note the API response includes your changes.  E.g., `wp-json/wc-analytics/reports/orders`.
3. If you modified an existing column or have filters in place for the reports table on the JS side, note the changes in your table.

### CODE REVIEW

**Comment by daigo75** [comment]
@joshuatf Thanks for the update. I'm trying to understand how this would work. Would that filter be triggered by every report, and the `$context` variable indicating which report is being generated? That is, it would become `orders`, `products`, etc, depending on the report?

**Comment by joshuatf** [comment]
>That is, it would become orders, products, etc, depending on the report?

That's correct, @daigo75.  This filter would give you access to all of the columns, context (report name), and the database table name, allowing the columns to be modified or add additional columns.  There's an example above that demonstrates this.

Feel free to ping me with any questions you might have about this. 😄

**Comment by daigo75** [comment]
Thanks, this looks suitable. How do I build this new version for testing? Is there a branch I can check out and build?

**Comment by joshuatf** [comment]
Hi @daigo75, you can checkout this branch `add/4147` and then run `composer install` and `npm i && npm run build` to test this branch out.

After we get things approved here, we can also have this ready to use in the next version of WooCommerce Admin.

**Comment by timmyc** [comment]
@daigo75 just wondering if you have had a chance to test this out yet. Want to make sure this provides the proper filters you need in your extension prior to us shipping it.

**Comment by daigo75** [comment]
I didn't have the time to run the tests, yet, I will do that as soon as I can. Thanks for checking.

**Review by psealock — APPROVED** [review]
This certainly is a powerful filter, especially when things like `woocommerce_analytics_clauses_select_orders_subquery` don't suffice. Do you think it has potential for misuse? Is there any way to mitigate that?

This is testing well and working as expected.

**Comment by daigo75** [comment]
@psealock all filters and actions have potential for misuse. That's a conversation we already had when I first created the related issue. The idea of a filter is to allow 3rd parties to alter some element. They could replace it with utter garbage and bring the whole site down, but that's an inherent risk with any filter.

I don't think there is any need, nor possibility to mitigate the chances of misuse. If you tried to restrict what the filter can return, that would limit the filter's usefulness and introduce a potential maintenance nightmare as well on your side. For example, in my case the returned columns would be the same original ones, but generated differently (i.e. in the way needed to prepare a correct report). Short of building and running the while SQL to check if the columns are there, there would be no way to know if the output returned by a filter makes sense. 

The bottom line is that, once a developer uses a filter, the risk of such filter breaking something will always be there. That's just how things are. ☺

**Comment by joshuatf** [comment]
>Do you think it has potential for misuse? Is there any way to mitigate that?

I do think that there is some potential for misuse, but hopefully no more than already exists.  We have filters for adding in column data prior to this, so this doesn't expose further risk there.  On the database side, we are granting access to column data, but I believe this to be mostly a matter of convenience as this could be done outside if this hook if someone wanted.

>The bottom line is that, once a developer uses a filter, the risk of such filter breaking something will always be there. That's just how things are. ☺

Overall, I agree with this sentiment and we generally want to err on the side of extensibility.  But there are definitely filters and hooks that have more exposure and are more likely to be abused, so @psealock's question is a good one to double check before merging something like this in.

**Comment by daigo75** [comment]
@josuhatf perhaps it's just a different perception. I rarely object to adding a filter when it can make a solution more extensible, as that's where a lot of the "WordPress power" comes from. Of course, I wouldn't put filters everywhere, because processing them adds some overhead, but, personally, I'm not too concerned about what other developers do with them. If they want to mess things up, they will find a way. I lost count of the sites where the code in WP core or in some other plugin was modified...

**Comment by daigo75** [comment]
@joshuatf I would like to confirm that the filter does exactly what I need. I'm now implementing the necessary code to extend the columns and running some tests.

Will the new version be part of WooCommerce 4.5?

**Comment by timmyc** [comment]
@daigo75 really glad this will allow you to add support to your extension for wc-admin! This will probably ship out in WooCommerce 4.6 - so our release in early October.
